### PR TITLE
Steps 5 & 6 of walkthrough: Datastore and additional features

### PR DESCRIPTION
**Pull request by jindaln**
Moved data to datastore; created three servlets for making new comments, listing out comments, and deleting comments. Added delete button in index.html and deleteMessages function in script.js. Additionally, created select element in index.html that allows users to select how many comments they want to see on the page.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DeleteCommentsServlet.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    Query query = new Query("Comment");
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+        datastore.delete(entity.getKey());
+    }
+
+    return;
```
Comment (v1shanker) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteCommentsServlet.java:33]: nit: you don't need return at the end of a void function.

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/list_comments")
+public class ListCommentsServlet extends HttpServlet{
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        Query query = new Query("Comment");
+
```
Comment (v1shanker) [inline portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:25]: nit: remove extra whitespace

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withLimit(Integer.parseInt(request.getParameter("max_comments"))));
```
Comment (v1shanker) [inline portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:27]: this is a really long line. if a line goes past 100 characters, we wrap the line. if you are using chaining (like you do here), the . is a good place to break it.

List<Entity> results = datastore.prepare(query).
asList(FetchOptions.Builder.withLimit(
  Integer.parseInt(request.getParameter("max_comments")
)));

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+        List<Comment> comments = new ArrayList<>();
+        for (Entity entity : results) {
+            long id = entity.getKey().getId();
+            String name = (String) entity.getProperty("name");
+            String comment = (String) entity.getProperty("comment");
+
+            Comment new_comment = new Comment(id, name, comment);
```
Comment (v1shanker) [inline portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:35]: camelCase

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            comments.add(new_comment);
+        }
+        Gson gson = new Gson();
+        String json = gson.toJson(comments);
```
Comment (v1shanker) [inline portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:39]: you can combine these lines:

new Gson().toJson(comments);

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/new_comment")
+public class NewCommentServlet extends HttpServlet{
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response){
+        String name = request.getParameter("name");
```
Comment (v1shanker) [inline portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:22]: make "name", etc constants like in the previous diffs

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        taskEntity.setProperty("name", name);
+        taskEntity.setProperty("comment", comment);
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        datastore.put(taskEntity);
+
+        try {
+            response.sendRedirect("/index.html");
+        } catch (IOException e) {
+            System.out.println("Could not redirect");
```
Comment (v1shanker) [inline portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:35]: does this System.out.println go anywhere useful?

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -28,21 +28,37 @@ function addRandomGreeting() {
...
-        commentsElement.innerHTML = '';
-        commentsElement.appendChild(
-            createListElement('First Message: ' + comments[0]));
-        commentsElement.appendChild(
-            createListElement('Second Message: ' + comments[1]));
-        commentsElement.appendChild(
-            createListElement('Third Message: ' + comments[2]));
+    var max_comment = document.getElementById("max_comments").value;
+    fetch(`/list_comments?max_comments=${max_comment}`).then(response => response.json()).then((history) => {
+        const historyEl = document.getElementById('history');
```
Comment (v1shanker) [inline portfolio/src/main/webapp/script.js:33]: historyElement?

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -28,21 +28,37 @@ function addRandomGreeting() {
...
+  var name = document.forms["comments_form"]["name"].value;
+  var comment = document.forms["comments_form"]["comment"].value;
+  if (name == "") {
+    alert("Name must be filled out");
+    return false;
+  }
+  if (comment == "") {
+    alert("Comment must be filled out");
+    return false;
+  }
```
Comment (v1shanker) [inline portfolio/src/main/webapp/script.js:63]: return true at the end of the function?

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
+package com.google.sps.servlets;
```
Comment (jynCoder) [inline portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:1]: Nice code. :)

Comment (jindaln) [reply portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:1]: thanks :)

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java (hunk 7 above)
Comment (jindaln) [reply portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:35]: Changed to logger log4j2

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
+package com.google.sps.data;
+
+public class Comment{
+
```
Comment (v1shanker) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:4]: get rid of extra blank line

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+import com.google.appengine.api.datastore.Entity;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/list_comments")
+public class ListCommentsServlet extends HttpServlet{
+
```
Comment (v1shanker) [inline portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:21]: get rid of extra new line

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java (hunk 2 above)
Comment (v1shanker) [reply portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:27]: this wasn't addressed?

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+            Integer.parseInt(request.getParameter("max_comments"))));
+
+        List<Comment> comments = new ArrayList<>();
+        for (Entity entity : results) {
+            long id = entity.getKey().getId();
+            String name = (String) entity.getProperty(NAME);
+            String comment = (String) entity.getProperty(COMMENT);
+
+            Comment newComment = new Comment(id, name, comment);
+            comments.add(newComment);
```
Comment (v1shanker) [inline portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:41]: comments.add(new Comment(id, name, comment));

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+        List<Entity> results = datastore.prepare(query).
+            asList(FetchOptions.Builder.withLimit(
+            Integer.parseInt(request.getParameter("max_comments"))));
+
+        List<Comment> comments = new ArrayList<>();
+        for (Entity entity : results) {
+            long id = entity.getKey().getId();
+            String name = (String) entity.getProperty(NAME);
+            String comment = (String) entity.getProperty(COMMENT);
+
```
Comment (v1shanker) [inline portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:39]: no need for extra blank line

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java (hunk 5 above)
Comment (v1shanker) [reply portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:44]: this wasn't addressed?

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/new_comment")
+public class NewCommentServlet extends HttpServlet{
+
```
Comment (v1shanker) [inline portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:21]: extra new line

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java (hunk 2 above)
Comment (jindaln) [reply portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:27]: sorry

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java (hunk 5 above)
Comment (jindaln) [reply portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:44]: sorry not sure how it skipped me

File: portfolio/src/main/webapp/script.js [hunk 16]
```diff
@@ -28,21 +28,38 @@ function addRandomGreeting() {
...
 }
+
+function validateForm() {
+  var name = document.forms["comments_form"]["name"].value;
+  var comment = document.forms["comments_form"]["comment"].value;
+  if (name == "") {
+    alert("Name must be filled out");
+    return false;
+  }
+  if (comment == "") {
```
Comment (v1shanker) [inline portfolio/src/main/webapp/script.js:60]: == in javascript is weird. use === or !name (same for comment below).

**Pull request merged**